streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -146,96 +165,203 @@ public void processInOrder(final K key, final V value, final long timestamp) {
+                    Math.max(0, inputRecordTimestamp - 2 * windows.timeDifferenceMs()),
                     // to catch the current record's right window, if it exists, without more calls to the store
-                    timestamp + 1)
+                    inputRecordTimestamp + 1)
             ) {
                 while (iterator.hasNext()) {
-                    final KeyValue<Windowed<K>, ValueAndTimestamp<Agg>> next = iterator.next();
-                    windowStartTimes.add(next.key.window().start());
-                    final long startTime = next.key.window().start();
+                    final KeyValue<Windowed<K>, ValueAndTimestamp<Agg>> windowBeingProcessed = iterator.next();
+                    final long startTime = windowBeingProcessed.key.window().start();
+                    windowStartTimes.add(startTime);
                     final long endTime = startTime + windows.timeDifferenceMs();
+                    final long windowMaxRecordTimestamp = windowBeingProcessed.value.timestamp();
 
-                    if (endTime < timestamp) {
-                        leftWinAgg = next.value;
-                        if (isLeftWindow(next)) {
-                            latestLeftTypeWindow = next.key.window();
-                        }
-                    } else if (endTime == timestamp) {
+                    if (endTime < inputRecordTimestamp) {
+                        leftWinAgg = windowBeingProcessed.value;
+                        previousRecordTimestamp = windowMaxRecordTimestamp;
+                    } else if (endTime == inputRecordTimestamp) {
                         leftWinAlreadyCreated = true;
-                        putAndForward(next.key.window(), next.value, key, value, closeTime, timestamp);
-                    } else if (endTime > timestamp && startTime <= timestamp) {
-                        rightWinAgg = next.value;
-                        putAndForward(next.key.window(), next.value, key, value, closeTime, timestamp);
-                    } else {
+                        if (windowMaxRecordTimestamp < inputRecordTimestamp) {
+                            previousRecordTimestamp = windowMaxRecordTimestamp;
+                        }
+                        updateWindowAndForward(windowBeingProcessed.key.window(), windowBeingProcessed.value, key, value, closeTime, inputRecordTimestamp);
+                    } else if (endTime > inputRecordTimestamp && startTime <= inputRecordTimestamp) {
+                        rightWinAgg = windowBeingProcessed.value;
+                        updateWindowAndForward(windowBeingProcessed.key.window(), windowBeingProcessed.value, key, value, closeTime, inputRecordTimestamp);
+                    } else if (startTime == inputRecordTimestamp + 1) {
                         rightWinAlreadyCreated = true;
+                    } else {
+                        throw new IllegalStateException("Unexpected window found when processing sliding windows");
```

Comment:
nit: log an error and include the relevant info (eg `windowStart` and `inputRecordTimestamp` at least). Same for the IllegalStateException in `processEarly`